src/script.c

```diff
@@ -108,10 +108,70 @@ int scriptInterrupt(scriptRunCtx *run_ctx) {
-void scriptPrepareForRun(scriptRunCtx *run_ctx, client *engine_client, client *caller, const char *funcname) {
+int scriptPrepareForRun(scriptRunCtx *run_ctx, client *engine_client, client *caller, const char *funcname, uint64_t script_flags, int ro) {
     serverAssert(!curr_run_ctx);
-    /* set the curr_run_ctx so we can use it to kill the script if needed */
-    curr_run_ctx = run_ctx;
+
+    int running_stale = server.masterhost &&
+            server.repl_state != REPL_STATE_CONNECTED &&
+            server.repl_serve_stale_data == 0;
+
+    if (!(script_flags & SCRIPT_FLAG_EVAL_COMPAT_MODE)) {
+        if ((script_flags & SCRIPT_FLAG_NO_CLUSTER) && server.cluster_enabled) {
+            addReplyError(caller, "Can not run script on cluster, 'no-cluster' flag is set.");
+            return C_ERR;
+        }
+
+        if (!(script_flags & SCRIPT_FLAG_ALLOW_OOM) && server.script_oom && server.maxmemory) {
+            addReplyError(caller, "-OOM allow-oom flag is not set on the script, "
+                                  "can not run it when used memory > 'maxmemory'");
+            return C_ERR;
+        }
+
+        if (running_stale && !(script_flags & SCRIPT_FLAG_ALLOW_STALE)) {
+            addReplyError(caller, "-MASTERDOWN Link with MASTER is down, "
+                             "replica-serve-stale-data is set to 'no' "
+                             "and 'allow-stale' flag is not set on the script.");
+            return C_ERR;
+        }
+
+        if (!(script_flags & SCRIPT_FLAG_NO_WRITES)) {
+            /* Script may perform writes we need to verify:
+             * 1. we are not a readonly replica
+             * 2. no disk error detected
+             * 3. command is not `fcall_ro`/`eval[sha]_ro` */
+            if (server.masterhost && server.repl_slave_ro && caller->id != CLIENT_ID_AOF
+                && !(caller->flags & CLIENT_MASTER))
+            {
+                addReplyError(caller, "Can not run script with write flag on readonly replica");
+                return C_ERR;
+            }
+
+            int deny_write_type = writeCommandsDeniedByDiskError();
+            if (deny_write_type != DISK_ERROR_TYPE_NONE && server.masterhost == NULL) {
+                if (deny_write_type == DISK_ERROR_TYPE_RDB)
+                    addReplyError(caller, "-MISCONF Redis is configured to save RDB snapshots, "
+                                     "but it is currently not able to persist on disk. "
+                                     "So its impossible to run scripts that have 'write' flag on.");
+                else
+                    addReplyErrorFormat(caller, "-MISCONF Redis is configured to persist data to AOF, "
+                                           "but it is currently not able to persist on disk. "
+                                           "So its impossible to run scripts that have 'write' flag on. "
+                                           "AOF error: %s", strerror(server.aof_last_write_errno));
```

Comment:
Aren't we exposing too much information to the client here? I think in general we should consider shorter messages here, as in most cases it's going to be an app rather than a user on the other end, hopefully just parsing the error code and throwing the right exception type.

Reply:
Originally (in the function flags PR), i asked that this error message will be similar to the one in server.c.
the extra part about script flags may be redundant, but that may depend on the default of the flag.

i.e. if the default was to deny writes, then adding a message that suggests to add `write` flag could save someone a lot of effort searching for the problem.

specifically about this message, it's outdated since we changed the `write` flag to `no-writes`.
so
1. we must fix the message
2. since the default for this script is permissive (even read only scripts will be blocked), i think we should keep it.

Reply:
@MeirShpilraien wdyt, It's your error message I think.

Reply:
fixing comment to mention `no-writes`.
regarding returning the last aof error, it's been there for years, see fe8352540fa5d6157648427b0651de9d5574e48d. So I'm inclined to keeping it.